network/phonebook.go

```diff
@@ -114,24 +114,36 @@ func (p *ThreadsafePhonebook) ReplacePeerList(they []string) {
 	copy(p.addrs, they)
 }
 
-// MultiPhonebook contains several phonebooks
+// MultiPhonebook contains a map of phonebooks
 type MultiPhonebook struct {
-	phonebooks []Phonebook
-	lock       deadlock.RWMutex
+	phonebookMap map[string]*Phonebook
+	lock         deadlock.RWMutex
+}
+
+// MakeMultiPhonebook constructs and returns a new Multi Phonebook
+func MakeMultiPhonebook() *MultiPhonebook {
+	return &MultiPhonebook{phonebookMap: make(map[string]*Phonebook)}
 }
 
 // GetAddresses returns up to N address
 // TODO: this implementation does a bunch of extra copying, make it more efficient
 func (mp *MultiPhonebook) GetAddresses(n int) []string {
 	mp.lock.RLock()
 	defer mp.lock.RUnlock()
-	if len(mp.phonebooks) == 1 {
-		return mp.phonebooks[0].GetAddresses(n)
+
+	phonebookList := []Phonebook{}
+
+	for _, phonebook := range mp.phonebookMap {
+		phonebookList = append(phonebookList, *phonebook)
 	}
-	sizes := make([]int, len(mp.phonebooks))
+
+	if len(phonebookList) == 1 {
```

Comment:
Change for loop: `if len(mp.phonebookMap) == 0 { return phonebook.GetAddresses(n) }`

Reply:
There is no longer a phonebook variable to get addresses from, only the map which includes a "default" map.

Reply:
The for loop has a phonebook iterator - that's what I meant.

Reply:
Updated to iterate over the map.